### PR TITLE
Bania

### PR DESCRIPTION
**Pull request by BaniaFonseca**
Hi guys, were are in a hurry trying to finishing the deployment pipeline ...

I just added a directory named build, this directory is to save the model that will be deployed and I already put some demo models on the directory .

I am working on a script that download the model to be deployed from build directory, we a doing this to make it easy to connect the two pipeline: model development and model deployment

### CODE REVIEW

File: gym/eval.sh [hunk 1]
```diff
@@ -0,0 +1,2 @@
+chmod 777 -Rf tf_od_api/workspace/train_core/
+nvidia-docker run -it -v $(pwd)/tf_od_api/workspace/train_core:/home registry.fsoc.hpi.uni-potsdam.de/coral-ai > tf_od_api/workspace/train_core/eval.log
```
Comment (tjiagoM) [inline gym/eval.sh:2]: You run something on `$(pwd)/tf_od_api/...` but then you redirect the output to only `tf_od_api/...`. You should probably have the `pwd` on both sides

File: gym/exporter_main_v2.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+r"""Tool to export an object detection model for inference.
+
+Prepares an object detection tensorflow graph for inference using model
+configuration and a trained checkpoint. Outputs associated checkpoint files,
+a SavedModel, and a copy of the model config.
+
+The inference graph contains one of three input nodes depending on the user
+specified option.
+  * `image_tensor`: Accepts a uint8 4-D tensor of shape [1, None, None, 3]
```
Comment (tjiagoM) [inline gym/exporter_main_v2.py:25]: Something odd on the documentation. You mention a "uint8" but the options you define on line 108 are strings.

File: gym/eval.sh (hunk 1 above)
Comment (tjiagoM) [inline gym/eval.sh:2]: You run something on `$(pwd)/tf_od_api/...` but then you redirect the output to only `tf_od_api/...`. You should probably have the `pwd` on both sides

File: gym/tf_od_api/workspace/train_core/core.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
+cd /opt/tf_od_api/workspace/train_core
+
+echo "Evaluating the model"
+
+python model_main_tf2.py \
+--model_dir=/home/models/ssd_mobilenet_v2_fpnlite \
```
Comment (tjiagoM) [inline gym/tf_od_api/workspace/train_core/core.sh:6]: I would say we might end up having our models in a folder other than `/home/...`. So, I'd put this as a variable at the beginning of the scrip to change if this folder changes in the future.

File: gym/tf_od_api/workspace/train_core/eval.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+cd /opt/tf_od_api/workspace/train_core
+
+echo "Evaluating the model"
+
+python model_main_tf2.py \
```
Comment (tjiagoM) [inline gym/tf_od_api/workspace/train_core/eval.sh:5]: Just like in the `train_core/core.sh`, I think we can define a variable at the beginning of the script with the exact folder we are using to store the models

File: gym/tf_od_api/workspace/train_core/model_main.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Binary to run train and evaluation on object detection model."""
+
+from __future__ import absolute_import
```
Comment (tjiagoM) [inline gym/tf_od_api/workspace/train_core/model_main.py:17]: If we are using python3, we should probably remove these future statements... They are usually useful for backwards compatibility, but we probably want to stick with a bunch of specific versions so we are sure we have exactly the same environment across everything. We can leverage miniconda when defining the docker file. For example, your Dockerfile right now doesn't have any specification on which python version it's using (ie it just uses "python").


In my opinion, you should generate a .yaml file from your current anaconda environment:
https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#exporting-the-environment-yml-file

And then, when you create the anaconda environment in the docker you specify that .yaml file:
https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-from-an-environment-yml-file

This will also save us in the case of tf being uploaded and breaking compatibility: instead of installing tensorflow-gpu, you specify which version you really want.



I'm guessing you can also use miniconda rather than anaconda for the docker, to have a lighter one:
https://hub.docker.com/r/continuumio/miniconda

File: gym/tf_od_api/workspace/train_core/pre-trained-models/efficientdet_d0/pipeline.config [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+      localization_weight: 1.0
+    }
+    encode_background_as_zeros: true
+    normalize_loc_loss_by_codesize: true
+    inplace_batchnorm_update: true
+    freeze_batchnorm: false
+    add_background_class: false
+  }
+}
+train_config {
```
Comment (tjiagoM) [inline gym/tf_od_api/workspace/train_core/pre-trained-models/efficientdet_d0/pipeline.config:133]: we need to keep in mind we will have to add more data augmentation options in the future, but I guess we can leave it like that for now. 

Also, couldn't `random_horizontal_flip` and `random_scale_crop_and_pad_to_square` all be inside the same `data_augmentation_options` rather than in two separate ones?

File: gym/tf_od_api/workspace/train_core/train.sh.old [hunk 7]
```diff
@@ -0,0 +1,34 @@
+cd /opt/tf_od_api/workspace/train_core
```
Comment (tjiagoM) [inline gym/tf_od_api/workspace/train_core/train.sh.old:1]: For consistency, we probably want the `$(pwd)` here as well (before the `tf_od_api/`). And I maintain the same suggestion here to define a global variable for the models location.

File: gym/train.sh [hunk 8]
```diff
@@ -0,0 +1,3 @@
+chmod 777 -Rf tf_od_api/workspace/train_core/
+nvidia-docker run -it -v $(pwd)/tf_od_api/workspace/train_core:/home registry.fsoc.hpi.uni-potsdam.de/coral-ai > tf_od_api/workspace/train_core/train.log
```
Comment (tjiagoM) [inline gym/train.sh:2]: Just like in `eval.sh`, you probably want to have `$(pwd)` in the output directory as well

**Review by tjiagoM — CHANGES_REQUESTED** [review]
Quite a big PR this one :D I've left a few comments and questions throughout some files that I think you can address before merging this. 

Thanks, good work!

**Comment by tjiagoM** [comment]
Also, I forgot to mention, inside the `build/` directory you have many things (eg models and labels) related to coco dataset, but weren't you using things based on the Aquarium dataset?

File: gym/tf_od_api/workspace/train_core/pre-trained-models/efficientdet_d0/pipeline.config [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+  use_moving_averages: false
+  batch_size: 1;
+}
+
+eval_input_reader: {
+  label_map_path: "PATH_TO_BE_CONFIGURED/label_map.txt"
+  shuffle: false
+  num_epochs: 1
+  tf_record_input_reader {
+    input_path: "PATH_TO_BE_CONFIGURED/val2017-?????-of-00032.tfrecord"
```
Comment (tjiagoM) [inline gym/tf_od_api/workspace/train_core/pre-trained-models/efficientdet_d0/pipeline.config:189]: Are we leaving these `PATH_TO_BE_CONFIGURED` like this willingly, or should we change/remove these ones for clarity?

**Review by tjiagoM — CHANGES_REQUESTED** [review]

File: gym/tf_od_api/workspace/train_core/model_main.py (hunk 5 above)
Comment (BaniaFonseca) [reply gym/tf_od_api/workspace/train_core/model_main.py:17]: We won't need that module any longer, I will remove it

File: gym/exporter_main_v2.py (hunk 2 above)
Comment (BaniaFonseca) [reply gym/exporter_main_v2.py:25]: This module is part of tensorflow api for object  detection,  I didn't implement it myself  ... 

And about  the line 108 is defining command  lines arguments  

https://abseil.io/docs/python/guides/flags

File: gym/tf_od_api/workspace/train_core/pre-trained-models/efficientdet_d0/pipeline.config (hunk 9 above)
Comment (BaniaFonseca) [reply gym/tf_od_api/workspace/train_core/pre-trained-models/efficientdet_d0/pipeline.config:189]: I've just added  this pretrained  model, I didn't start work  with it  yet ... and about all other  comments and suggestions I can't  implement  them before  finishing the deployment  pipeline  first ... the model  deployment  pipeline  Indeed  needs improvements ... I will address  this issues as soon as possible  , thanks for commenting

File: gym/exporter_main_v2.py (hunk 2 above)
Comment (tjiagoM) [reply gym/exporter_main_v2.py:25]: Ok, if it was not implemented by you then it's not our problem, I know about the flags, my question is about misalignment between the code (line 108) and the documentation at this line. I guess maybe they have changed the code at a certain point in time but forgot to change the documentation. 

Actually, now that I see again, this is even more different than I though. Ok, nevermind, let's move on on this.

File: gym/tf_od_api/workspace/train_core/pre-trained-models/efficientdet_d0/pipeline.config (hunk 9 above)
Comment (tjiagoM) [reply gym/tf_od_api/workspace/train_core/pre-trained-models/efficientdet_d0/pipeline.config:189]: Ok, just wanted to confirm. Let's leave model specifications for later, then :)

File: gym/tf_od_api/workspace/train_core/model_main.py (hunk 5 above)
Comment (tjiagoM) [reply gym/tf_od_api/workspace/train_core/model_main.py:17]: Ok, cool. Specifications of module dependencies using a .yml file is probably important already for this PR as well. I think it helps in having the deployment more organised.

File: convert_to_tflite.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import tensorflow as tf
```
Comment (nlauchande) [inline convert_to_tflite.py:1]: It's starting to look like a CLI. In case you want to try one, you can look at https://github.com/google/python-fire

File: gym/export_model.sh [hunk 11]
```diff
@@ -0,0 +1,15 @@
+echo "Exporting core model"
```
Comment (nlauchande) [inline gym/export_model.sh:1]: @BaniaFonseca @tjiagoM : Going back to the MLFlow applicability. This would be in an ideal scenario the right place to add it as a Model Registry. If we had one accessible from somewhere.

So the API side would just need to retrieve the model from a registry with interrelated metadata and versioning information.

My initial biggest concern is that we needed hosting, I am currently trialling we can use the community edition self hosted version.

File: .gitignore [hunk 12]
```diff
@@ -8,7 +8,6 @@ __pycache__/
 
 # Distribution / packaging
```
Comment (nlauchande) [inline .gitignore:9]: I think we need to add some of the model outputs to the registry. I tried to pull this repo is at a GBs level.

File: gym/eval.sh (hunk 1 above)
Comment (BaniaFonseca) [reply gym/eval.sh:2]: We wont use this file any longer

File: convert_to_tflite.py (hunk 10 above)
Comment (BaniaFonseca) [reply convert_to_tflite.py]: thanks for the suggestion, but I am going to use "absl"

File: gym/export_model.sh (hunk 11 above)
Comment (BaniaFonseca) [reply gym/export_model.sh]: Maybe you can this out in a new branch, because I did not get it ...

File: gym/eval.sh (hunk 1 above)
Comment (tjiagoM) [reply gym/eval.sh:2]: Ah, ok, in that case `eval.sh` can just be removed from the repo I guess?

File: gym/tf_od_api/workspace/train_core/core.sh (hunk 3 above)
Comment (BaniaFonseca) [reply gym/tf_od_api/workspace/train_core/core.sh]: the pipeline has been changed ...
about the /home, it refer to /home inside the docker container, I can not explain it here ...
but the idea is: we've mount and bind our workspace directory to /home of our container, doing so /home is like a pointer to our workspace

File: gym/tf_od_api/workspace/train_core/eval.sh (hunk 4 above)
Comment (BaniaFonseca) [reply gym/tf_od_api/workspace/train_core/eval.sh]: this has been solved in the recent commit

File: gym/tf_od_api/workspace/train_core/model_main.py (hunk 5 above)
Comment (BaniaFonseca) [reply gym/tf_od_api/workspace/train_core/model_main.py]: for now I am using Dockerfile, when the pipeline is more stable, then maybe can move to .yml

File: gym/tf_od_api/workspace/train_core/pre-trained-models/efficientdet_d0/pipeline.config (hunk 6 above)
Comment (BaniaFonseca) [reply gym/tf_od_api/workspace/train_core/pre-trained-models/efficientdet_d0/pipeline.config]: I can not answer this one

File: gym/tf_od_api/workspace/train_core/train.sh.old (hunk 7 above)
Comment (BaniaFonseca) [reply gym/tf_od_api/workspace/train_core/train.sh.old]: there are paths that a meant to be relative, but about the model locations it is not a problem they are all stored on gym/ai4coral

File: gym/train.sh (hunk 8 above)
Comment (BaniaFonseca) [reply gym/train.sh]: the redirection  has been removed, we only used $(pwd) because docker require it when mounting and binding a directory

**Review by tjiagoM — APPROVED** [review]
I've checked the comments and new changes. This PR is getting too big with many different points being discussed in the same place so it's probably a good time to merge. Very nice work! :)

P.S.: While reviewing, I noticed that the "Files changed" tab specifies the size of each file, and as Natu mentioned in the meeting we have some big ones around. This will be solved next time :)

File: .gitignore (hunk 12 above)
Comment (tjiagoM) [reply .gitignore:9]: Yes, I agree! We talked about that in our meeting, but I think we can accept the PR as it is now, and leave that file reduction for another PR maybe.
